### ISSUE TITLE
Refactor training hyperparameters in `SetFitTrainer` to dedicated class

### ISSUE THREAD

**Issue by lewtun** [huggingface/setfit#179]
The current implementation of `SetFitTrainer` has become a bit bloated because it accepts all the hyperparameters for training as separate arguments. In hindsight, it would have been better to collect all the hyperparameters in a separate data class (similar to the [`TrainingArguments`](https://huggingface.co/docs/transformers/v4.24.0/en/main_classes/trainer#transformers.TrainingArguments) of `transformers`). At the same time, it doesn't seem necessary to prepend `SetFit` to the name of these classes.

To that end, it would be great to:

- [ ] Create new `TrainingArguments` and `Trainer` classes, where `Trainer` accepts all the hyperparameters as an `args` parameter.
- [ ] Add a deprecation notice on `SetFitTrainer` to inform users about the new `Trainer` (and let them know it will be removed in 2 minor releases time)

**Comment by tomaarsen** [comment]
I'll get working on this in a week on the 21st, if it hasn't been claimed by then!

- Tom Aarsen

**Comment by lewtun** [comment]
Thanks @tomaarsen - that would be wonderful! Feel free to ping me if you need any guidance :)

**Comment by tomaarsen** [comment]
Now that `DistillationSetFitTrainer` has entered the picture, how do we want to structure the new classes?
`Trainer`, `DistillationTrainer` and `TrainingArguments`?

Furthermore, I noticed that training using `use_differentiable_head=True` requires some freezing/unfreezing and multiple calls to `trainer.train()`. My intuition for a `Trainer` is that it requires one call to perform all of the training, which is not what we're seeing now. I also noticed that the differentiable head training does not allow flexibility in all parameters, like the optimizer, the scheduler, and its parameters. 

If we indeed want to allow one call to train the model fully, then we may want to create the following structure:

```python
@dataclass
class TrainingArguments:
    body_learning_rate = 2e-5
    num_iterations = 20
    num_epochs = 1
    batch_size = 16
    l2_weight = 0.0
    grad_accum_steps = 4
    loss_class = losses.CosineSimilarityLoss
    scheduler: str = "WarmupLinear"
    optimizer_class = AdamW
    ...


class BodyTrainingArguments(TrainingArguments):
    pass


class HeadTrainingArguments(TrainingArguments):
    loss_class = losses.CosineSimilarityLoss
    body_learning_rate = 1e-5
    head_learning_rate = 1e-2
    num_epochs = 25
    freeze_body = False
    freeze_head = False
    ...
```
```python
class Trainer:
    def __init__(
        self,
        model: Optional[SetFitModel] = None,
        body_args: Optional[BodyTrainingArguments] = None,
        head_args: Optional[HeadTrainingArguments] = None,
        train_dataset: Dataset = None,
        eval_dataset: Dataset = None,
        model_init: Optional[Callable[[], SetFitModel]] = None,
    ) -> None:
        if head_args and isinstance(model.model_head, LogisticRegression):
            # Something along the lines of:
            warnings.warn(
                "head_args will be ignored when the model head is LogisticRegression"
            )

        ...

    def train(
        self,
        body_args: Optional[BodyTrainingArguments] = None,
        head_args: Optional[HeadTrainingArguments] = None,
    ):
        # These parameters may be removed to avoid confusion with multiple places to supply training arguments.
        self.train_body(body_args)
        self.train_head(head_args)

    def train_body(self, body_args: Optional[BodyTrainingArguments] = None):
        ...

    def train_head(self, head_args: Optional[HeadTrainingArguments] = None):
        ...


class DistillationTrainer(Trainer):
    def __init__(
        self,
        teacher_model: SetFitModel,
        student_model: Optional[SetFitModel] = None,
        body_args: Optional[BodyTrainingArguments] = None,
        head_args: Optional[HeadTrainingArguments] = None,
        train_dataset: Dataset = None,
        eval_dataset: Dataset = None,
        model_init: Optional[Callable[[], SetFitModel]] = None,
    ) -> None:
        pass
```

This includes `Trainer.train_body` and `Trainer.train_head` methods, which are used sequentially in `Trainer.train`. Furthermore, the `Trainer` class initialization and `Trainer.train` would accept 2 kinds of `TrainingArguments`: one for the body, and one for the head. These can be equipped with reasonable default options when no parameters are given. We also need to discuss whether it is better to only allow passing these args at initialization, rather than also at `train`/`train_body`/`train_head`.

I believe that this approach with two sets of arguments is preferable over the current situations, as e.g. `num_epochs` really seems like it should not be the same for both the body and the head, based on @blakechi's experimentations.

Alternatively, we can use one single `TrainingArguments` that simply uses many very specific parameters like `head_optimizer_class`, but I'm not a big fan of this. Note that if `use_differentiable_head=True` is supplied at `from_pretrained`, then `HeadTrainingArguments` will be fully ignored, and `BodyTrainingArguments` will be just like the `TrainingArguments` that we're all familiar with from other huggingface works and PyTorch Lightning.

It'll all be a bit tricky, as these parameters need to interface with SentenceTransformer, LogisticRegression and SetFitHead.

Would love to hear your thoughts.
cc: @lewtun 
cc: @blakechi for your familiarity with the differentiable head
cc: @orenpereg for your familiarity with the distillation trainer

- Tom Aarsen

**Comment by tomaarsen** [comment]
I'd like to start working on this once there is a consensus on whether this approach seems best!

**Comment by tomaarsen** [comment]
I'll wait on this for a bit due to the large amount of ready-to-go PRs that would inevitably cause large merge conflicts if I started already.

**Comment by tomaarsen** [comment]
Will be resolved by #439.
